File: cpp/src/day11.cc
#include "aoc.hh"
#include <cstdlib>
#include <unordered_set>

struct point {
    uint64_t y;
    uint64_t x;
};

static uint64_t solve(const std::string& input, uint64_t num_times);
static std::vector<uint64_t> row_expansions(const std::string& input);
static std::vector<uint64_t> col_expansions(const std::string& input);
static std::vector<point> find_galaxies(const std::string& input);
static uint64_t dist(point& a, point& b);

uint64_t day11p1(const std::string& input) { return solve(input, 1); }

uint64_t day11p2(const std::string& input) { return solve(input, 1000000); }

static uint64_t solve(const std::string& input, uint64_t num_times) {
    uint64_t res = 0;
    std::vector<uint64_t> row_exps = row_expansions(input);
    std::vector<uint64_t> col_exps = col_expansions(input);
    std::vector<point> galaxies = find_galaxies(input);
    size_t i, len = galaxies.size();
    for (i = 0; i < len; ++i) {
        point& p = galaxies[i];
        uint64_t new_y;
        uint64_t new_x;
        uint64_t row_exp = row_exps[p.y];
        uint64_t col_exp = col_exps[p.x];
        if (num_times == 1) {
            new_y = p.y + row_exp;
            new_x = p.x + col_exp;
        } else {
            new_y = p.y + (row_exp * (num_times - 1));
            new_x = p.x + (col_exp * (num_times - 1));
        }

        point new_point = {new_y, new_x};
        galaxies[i] = new_point;
    }

    for (i = 0; i < len; ++i) {
        uint64_t c = 0;
        point p = galaxies[i];
        size_t x;
        for (x = i + 1; x < len; ++x) {
            point pi = galaxies[x];
            uint64_t d = dist(p, pi);
            c += d;
        }
        res += c;
    }
    return res;
}

static std::vector<uint64_t> row_expansions(const std::string& input) {
    std::vector<uint64_t> res;
    std::vector<std::string> lines = split_str(input, '\n');
    uint64_t num_expansions = 0;
    for (auto& line : lines) {
        std::unordered_set<char> set;
        for (auto& c : line) {
            set.insert(c);
        }
        if (set.size() == 1) {
            num_expansions++;
        }
        res.push_back(num_expansions);
    }
    return res;
}

static std::vector<uint64_t> col_expansions(const std::string& input) {
    std::vector<uint64_t> res;
    std::vector<std::string> lines = split_str(input, '\n');
    size_t i, len = lines[0].length();
    uint64_t num_expansions = 0;
    for (i = 0; i < len; ++i) {
        std::unordered_set<char> set;
        for (auto& line : lines) {
            char ch = line[i];
            set.insert(ch);
        }
        if (set.size() == 1) {
            num_expansions++;
        }
        res.push_back(num_expansions);
    }
    return res;
}

static std::vector<point> find_galaxies(const std::string& input) {
    std::vector<point> res;
    std::vector<std::string> lines = split_str(input, '\n');
    size_t i, len = lines.size();
    size_t j, jlen = lines[0].length();
    for (i = 0; i < len; ++i) {
        std::string& line = lines[i];
        for (j = 0; j < jlen; ++j) {
            char ch = line[j];
            if (ch == '#') {
                point p = {i, j};
                res.push_back(p);
            }
        }
    }
    return res;
}

static uint64_t dist(point& a, point& b) {
    uint64_t y = abs((int)(a.y - b.y));
    uint64_t x = abs((int)(a.x - b.x));
    return y + x;
}

File: rs/src/day11.rs
pub fn part_one(input: &str) -> u64 {
    solve(input, 1)
}

pub fn part_two(input: &str) -> u64 {
    solve(input, 1_000_000)
}

fn solve(input: &str, num_times: u64) -> u64 {
    let mut res = 0;
    let mut galaxies = find_galaxies(input);
    let row_exps = row_expansions(input);
    let col_exps = col_expansions(input);

    let len = galaxies.len();

    for i in 0..len {
        let (y, x) = galaxies[i];
        let new_y: u64;
        let new_x: u64;
        let row_exp = row_exps[y as usize];
        let col_exp = col_exps[x as usize];
        if num_times == 1 {
            new_y = y + row_exp;
            new_x = x + col_exp;
        } else {
            new_y = y + (row_exp * (num_times - 1));
            new_x = x + (col_exp * (num_times - 1));
        }
        galaxies[i] = (new_y, new_x);
    }

    for i in 0..len {
        let mut c = 0;
        let (y, x) = galaxies[i];
        for j in (i + 1)..len {
            let (yi, xi) = galaxies[j];
            let d = dist((y, x), (yi, xi));
            c += d;
        }
        res += c;
    }
    res
}

fn row_expansions(input: &str) -> Vec<u64> {
    let mut res = Vec::new();
    let mut num_expansions = 0;
    for line in input.lines() {
        let mut s = std::collections::HashSet::new();
        for ch in line.chars() {
            s.insert(ch);
        }
        if s.len() == 1 {
            num_expansions += 1;
        }
        res.push(num_expansions);
    }
    res
}

fn col_expansions(input: &str) -> Vec<u64> {
    let mut res = Vec::new();
    let mut num_expansions = 0;
    let lines: Vec<&str> = input.lines().collect();
    let line_len = lines[0].len();
    for i in 0..line_len {
        let mut set = std::collections::HashSet::new();
        for line in lines.iter() {
            let ch = line.as_bytes()[i] as char;
            set.insert(ch);
        }
        if set.len() == 1 {
            num_expansions += 1;
        }
        res.push(num_expansions);
    }
    res
}

fn find_galaxies(input: &str) -> Vec<(u64, u64)> {
    let mut res = Vec::new();
    let lines = input.lines();
    for (i, line) in lines.enumerate() {
        for (j, ch) in line.chars().enumerate() {
            if ch == '#' {
                let p = (i as u64, j as u64);
                res.push(p);
            }
        }
    }
    res
}

fn dist((y1, x1): (u64, u64), (y2, x2): (u64, u64)) -> u64 {
    let y: u64 = ((y2 as i64 - y1 as i64) as i64).abs() as u64;
    let x: u64 = ((x2 as i64 - x1 as i64) as i64).abs() as u64;
    y + x
}
